widget/tree_test.go

```diff
@@ -0,0 +1,577 @@
+	})
+	t.Run("Initializer_Populated", func(t *testing.T) {
+		tree := &widget.Tree{
+			Children: func(uid string) (children []string) {
+				if uid == "" {
+					children = append(children, "a", "b", "c")
+				} else if uid == "c" {
+					children = append(children, "d", "e", "f")
+				}
+				return
+			},
+			IsBranch: func(uid string) bool {
+				return uid == "" || uid == "c"
+			},
+			CreateNode: func(branch bool) fyne.CanvasObject {
+				return &widget.Label{}
+			},
+			UpdateNode: func(uid string, branch bool, node fyne.CanvasObject) {
+				node.(*widget.Label).SetText(uid)
+			},
+		}
+		tree.OpenBranch("c")
+		var branches []string
+		var leaves []string
+		tree.Walk(func(uid string, branch bool, depth int) {
+			if branch {
+				branches = append(branches, uid)
+			} else {
+				leaves = append(leaves, uid)
+			}
+		})
+		assert.Equal(t, 2, len(branches))
+		assert.Equal(t, "", branches[0])
+		assert.Equal(t, "c", branches[1])
+		assert.Equal(t, 5, len(leaves))
+		assert.Equal(t, "a", leaves[0])
+		assert.Equal(t, "b", leaves[1])
+		assert.Equal(t, "d", leaves[2])
+		assert.Equal(t, "e", leaves[3])
+		assert.Equal(t, "f", leaves[4])
+	})
+	t.Run("NewTreeWithStrings", func(t *testing.T) {
+		data := make(map[string][]string)
+		internalwidget.AddTreePath(data, "foo", "foobar")
+		tree := widget.NewTreeWithStrings(data)
+		tree.OpenBranch("foo")
+		var branches []string
+		var leaves []string
+		tree.Walk(func(uid string, branch bool, depth int) {
+			if branch {
+				branches = append(branches, uid)
+			} else {
+				leaves = append(leaves, uid)
+			}
+		})
+		assert.Equal(t, 2, len(branches))
+		assert.Equal(t, "", branches[0]) // Root
+		assert.Equal(t, "foo", branches[1])
+		assert.Equal(t, 1, len(leaves))
+		assert.Equal(t, "foobar", leaves[0])
+	})
+	/* TODO Not currently possible as testDriver doesn't support listable URIs:
+	2020/08/31 16:49:18 Fyne error:  Unable to get lister for /var/folders/8n/1dd15fbn79s2w3l4x43v5c7w0000gn/T/test011812343
+	2020/08/31 16:49:18   Cause: test driver does support creating listable URIs yet
+	t.Run("NewTreeWithFiles", func(t *testing.T) {
+		tempDir, err := ioutil.TempDir("", "test")
+		assert.NoError(t, err)
+		defer os.RemoveAll(tempDir)
+		err = os.MkdirAll(path.Join(tempDir, "A"), os.ModePerm)
+		assert.NoError(t, err)
+		err = os.MkdirAll(path.Join(tempDir, "B"), os.ModePerm)
+		assert.NoError(t, err)
+		err = ioutil.WriteFile(path.Join(tempDir, "B", "C"), []byte("c"), os.ModePerm)
+		assert.NoError(t, err)
+
+		tree := widget.NewTreeWithFiles(storage.NewURI("file://" + tempDir))
+		tree.OpenAllBranches()
+		var branches []string
+		var leaves []string
+		tree.Walk(func(uid string, branch bool, depth int) {
+			if branch {
+				branches = append(branches, uid)
+			} else {
+				leaves = append(leaves, uid)
+			}
+		})
+		assert.Equal(t, 3, len(branches))
+		assert.Equal(t, tempDir, branches[0]) // Root
+		assert.Equal(t, filepath.Join(tempDir, "A"), branches[1])
```

Comment:
When merging the URI Child() feature is probably usable to do this.

Reply:
Sure, need #1108 to land first

Reply:
Done